src/caller.jl

```diff
@@ -0,0 +1,189 @@
+    :RING_CMD       => (libSingular.RING_CMD_CASTER,       false, ()),
+    :POLY_CMD       => (libSingular.POLY_CMD_CASTER,       true, ()),
+    :IDEAL_CMD      => (libSingular.IDEAL_CMD_CASTER,      true, ()),
+    :MODUL_CMD      => (libSingular.IDEAL_CMD_CASTER,      true, (:module,)),
+    :VECTOR_CMD     => (libSingular.POLY_CMD_CASTER,       true, (:vector,)),
+    :INT_CMD        => (libSingular.INT_CMD_CASTER,        false, ()),
+    :STRING_CMD     => (libSingular.STRING_CMD_CASTER,     false, ()),
+    :LIST_CMD       => (libSingular.LIST_CMD_TRAVERSAL,    false, ()),
+    :INTVEC_CMD     => (libSingular.INTVEC_CMD_CASTER,     false, ()),
+    :INTMAT_CMD     => (libSingular.INTMAT_CMD_CASTER,     false, ()),
+    :BIGINT_CMD     => (libSingular.BIGINT_CMD_CASTER,     false, ()),
+    :BIGINTMAT_CMD  => (libSingular.BIGINTMAT_CMD_CASTER,  false, ()),
+    :MAP_CMD        => (libSingular.MAP_CMD_CASTER,        false, ()),
+    :RESOLUTION_CMD => (libSingular.RESOLUTION_CMD_CASTER, true, (:resolution,)))
+
+casting_functions = nothing
+
+function create_casting_functions()
+    pair_array = Any[]
+    for (sym, func) in casting_functions_pre
+        push!(pair_array, mapping_types_reversed[sym] => func)
+    end
+    return Dict(pair_array...)
+end
+
+function convert_ring_content(value_list, rng)
+    return_dict = Dict{Symbol,Any}()
+    for i in value_list
+        return_dict[i[2]] = convert_return_value([false,i[3],i[1]], rng)
+    end
+    return return_dict
+end
+
+function convert_return_value(single_value, rng = nothing)
+    if single_value[1]
+        error("recieved list instead of single value")
+    end
+    cast = casting_functions[single_value[3]][1](single_value[2])
+    if cast isa Array{Any}
+        return recursive_translate(cast, rng)
+    elseif cast isa libSingular.ring
+        new_ring = rng(cast)
+        return [ new_ring, convert_ring_content(libSingular.get_ring_content(cast), new_ring) ]
+    elseif casting_functions[single_value[3]][2]
+        if length(casting_functions[single_value[3]][3]) > 0
+            cast = rng(cast, Val(casting_functions[single_value[3]][3][1]))
+        else
+            cast = rng(cast)
+        end
+    end
+    return cast
+end
+
+
+function convert_return_list(list_value, ring = nothing)
+    if list_value[1]
+        return map(i -> convert_return_value(i, ring), list_value[2:end])
+    end
+    return convert_return_value(list_value, ring)
+end
+
+function get_ring(arg_list)
+    ring = nothing
+    for i in arg_list
+        current_ptr = nothing
+        try
+            current_ptr = i.ptr
+        catch
+            continue
+        end
+        if current_ptr isa poly
+            return parent(i)
+        elseif current_ptr isa ideal
+            return parent(i).base_ring
+        end
+    end
+    return ring
+end
+
+function prepare_argument(x::Array{Int64,1})
+    return Any[ mapping_types_reversed[:INTVEC_CMD], libSingular.jl_array_to_intvec(x) ], nothing
```

Comment:
I find this whole business of generating an array in the C code in function `get_type_mapper `, exporting this to Julia, then turning it into a "reverse" mapping dictionary `mapping_types_reversed` there, then using that to convert a Julia symbol to an integer on the Julia level, then passing this back to the kernel, overall somewhat hard to track.

Can't we radically simplify this, by creating that "reverse mapping" dictionary in the kernel (possibly even as a plain C++ map!), and never exporting it from there? Then the above line would become just this:
```suggestion
    return Any[ :INTVEC_CMD, libSingular.jl_array_to_intvec(x) ], nothing
```
And in `caller.cpp` code, you only need to modify `translate_singular_type`, to expect a julia symbol as first argument, instead of integer. It then only needs to modify that integer.

I believe with this relatively straightforward change, things get quite a bit simpler.

Reply:
(I missed that also `create_casting_functions` and friends would need to be updated).

Anyway, thinking some more about this, it would probably be even simpler if indeed some of this stuff was replaced by a program that generates the required data (just as we do in https://github.com/gap-packages/SingularInterface) -- then that C/C++ program could in particular creates a `.jl` file which simply defines Julia constants with the values of `INTVEC_CMD` etc.. Then, we can use these everywhere instead of symbols, and completely dispense with the conversions between symbols and ints. Also, Julia could then optimize the code some more, I guess.

Reply:
Looking at the CxxWrap.jl "documentation" (i.e., its README -- is there anything else?), I guess one could also export the `yytokentype` enum to Julia. But from the explanations there, it is unclear to me whether one can convert the enum values into an `Cint`, as we would want.

Reply:
I think we should open a ticket for this and deal with it later. I doubt Sebastian has time to work on such a thing now.